cmd/postpolicyform.go

```diff
@@ -243,34 +259,66 @@ func checkPostPolicy(formValues http.Header, postPolicyForm PostPolicyForm) erro
 	condPassed := true
 
 	// Iterate over policy conditions and check them against received form fields
-	for cond, v := range postPolicyForm.Conditions.Policies {
-		// Form fields names are in canonical format, convert conditions names
-		// to canonical for simplification purpose, so `$key` will become `Key`
-		formCanonicalName := http.CanonicalHeaderKey(strings.TrimPrefix(cond, "$"))
-		// Operator for the current policy condition
-		op := v.Operator
-		// If the current policy condition is known
-		if startsWithSupported, condFound := startsWithConds[cond]; condFound {
-			// Check if the current condition supports starts-with operator
-			if op == policyCondStartsWith && !startsWithSupported {
-				return fmt.Errorf("Invalid according to Policy: Policy Condition failed")
-			}
-			// Check if current policy condition is satisfied
-			condPassed = checkPolicyCond(op, formValues.Get(formCanonicalName), v.Value)
-			if !condPassed {
-				return fmt.Errorf("Invalid according to Policy: Policy Condition failed")
-			}
-		} else {
-			// This covers all conditions X-Amz-Meta-* and X-Amz-*
-			if strings.HasPrefix(cond, "$x-amz-meta-") || strings.HasPrefix(cond, "$x-amz-") {
-				// Check if policy condition is satisfied
+	for _, policy := range postPolicyForm.Conditions.Policies {
+		for cond, v := range policy {
+			// Form fields names are in canonical format, convert conditions names
+			// to canonical for simplification purpose, so `$key` will become `Key`
+			formCanonicalName := http.CanonicalHeaderKey(strings.TrimPrefix(cond, "$"))
```

Comment:
We should not remove this prefix `$` for `startsWith` condition. This will fail for the correct case - If we actually try to upload a `key` prefixed object. Because, we fail here 

```
if _, condFound := startsWithConds[cond]; !condFound
```

as the keys in `startsWithConds` are prefixed with `$`